frontend/src/views/Game.tsx

```diff
@@ -1,45 +1,65 @@
 import React, { useEffect, useState } from 'react';
-import { useLocation } from 'react-router-dom';
+import { useHistory, useLocation } from 'react-router-dom';
 import SplitterLayout from 'react-splitter-layout';
+import { useBeforeunload } from 'react-beforeunload';
 import Editor from '../components/core/Editor';
-import { ErrorResponse } from '../api/Error';
+import { errorHandler, ErrorResponse } from '../api/Error';
 import { Problem, getProblems } from '../api/Problem';
 import { Room } from '../api/Room';
-import {
+import MainContainer, {
   FlexContainer, FlexInfoBar, Panel, SplitterContainer,
 } from '../components/core/Container';
 import ErrorMessage from '../components/core/Error';
 import { ProblemHeaderText, Text } from '../components/core/Text';
 import 'react-splitter-layout/lib/index.css';
 import { checkLocationState } from '../util/Utility';
+import Loading from '../components/core/Loading';
 
 type LocationState = {
   room: Room,
 }
 
 function GamePage() {
+  // Get history object to be able to move between different pages
+  const history = useHistory();
   const location = useLocation<LocationState>();
   const [room, setRoom] = useState<Room | null>(null);
   const [problems, setProblems] = useState<Problem[]>([]);
+  const [fullPageLoading, setFullPageLoading] = useState<boolean>(true);
   const [error, setError] = useState<string>('');
 
+  /**
+   * Display beforeUnload message to inform the user that they may lose
+   * their code / data if they leave the page.
+   * Some browsers will display this message, others will display a fixed
+   * message; see https://github.com/jacobbuck/react-beforeunload.
+   */
+  useBeforeunload(() => 'Leaving this page may cause you to lose your current code and data.');
+
   // Called every time location changes
   useEffect(() => {
     if (checkLocationState(location, 'room')) {
       setRoom(location.state.room);
+
+      // Get the game problem.
+      getProblems().then((res) => {
+        if (!(res as Problem[]).length) {
+          setError('Problem cannot be found.');
+        } else {
+          setProblems(res as Problem[]);
+          setError('');
+        }
+      }).catch((err) => {
+        setError((err as ErrorResponse).message);
+        setProblems([]);
+      });
+      setFullPageLoading(false);
```

Comment:
nit: I would move this to just under `setRoom(...)` so it's more clear it's part of the if statement and not the then/catch blocks

Reply:
or maybe it actually is supposed to be in the getProblems callback?

Reply:
I was hoping all of the page information could be populated before the actual page is shown; eventually, the loading symbol could be a prettier transition, but for now, I thought text would be a fine substitute.

I don't think it should be in the getProblems callback, since if the problems are not fetched properly, that should be displayed to the user rather than a static loading page.

Reply:
Sounds good - FYI I think right now the problem actually does load in a little bit later than the rest of the page (cause of async stuff), but personally I actually kind of like that so no changes here on my end unless that's something you care for 👍